File: masker/policy.py
"""Policy engine. CODEX OWNS THIS FILE.

Maps a DetectionResult to a PolicyDecision (route + policy + rationale).
Ships a HIPAA-first baseline so the pipeline runs end-to-end. Codex can
swap in a richer policy DSL — keep the signature stable.

Contract (see AGENTS.md):
    decide(detection: DetectionResult, *, policy_name: PolicyName) -> PolicyDecision
"""

from __future__ import annotations

from .contracts import DetectionResult, EntityType, PolicyDecision, PolicyName

_HIGH_RISK_LOCAL_ONLY: set[EntityType] = {EntityType.SSN, EntityType.MRN}


def decide(
    detection: DetectionResult,
    *,
    policy_name: PolicyName = "hipaa_base",
) -> PolicyDecision:
    """Decide the route for a turn given detected entities.

    Routes:
      - local-only: never leaves the device (highest sensitivity)
      - masked-send: forward to LLM with sensitive spans replaced
      - safe-to-send: forward verbatim (no PHI/PII detected)
    """
    types = {e.type for e in detection.entities}

    if policy_name == "hipaa_logging":
        if detection.has_sensitive:
            return PolicyDecision(
                route="masked-send",
                policy=policy_name,
                rationale="Strict logging policy: any sensitive data must be masked before traversal.",
            )

    if policy_name == "hipaa_clinical":
        if types & _HIGH_RISK_LOCAL_ONLY:
            return PolicyDecision(
                route="local-only",
                policy=policy_name,
                rationale="Direct identifiers (SSN/MRN) must stay on-device under clinical policy.",
            )
        if EntityType.HEALTH_CONTEXT in types and detection.entities:
            return PolicyDecision(
                route="masked-send",
                policy=policy_name,
                rationale="Clinical context with identifiers → mask identifiers, keep medical context.",
            )

    if types & _HIGH_RISK_LOCAL_ONLY:
        return PolicyDecision(
            route="local-only",
            policy=policy_name,
            rationale=f"High-risk identifiers detected: {sorted(t.value for t in types & _HIGH_RISK_LOCAL_ONLY)}",
        )

    if detection.has_sensitive:
        sensitive = sorted({e.type.value for e in detection.entities if e.type != EntityType.HEALTH_CONTEXT})
        return PolicyDecision(
            route="masked-send",
            policy=policy_name,
            rationale=f"Sensitive entities present: {sensitive}",
        )

    return PolicyDecision(
        route="safe-to-send",
        policy=policy_name,
        rationale="No sensitive entities detected.",
    )

File: platform/masker-core/crates/masker/src/policy.rs
//! Policy engine. Mirrors the HIPAA-first baseline in
//! `platform/masker-library/masker/policy.py`.

use std::collections::HashSet;

use crate::contracts::{DetectionResult, EntityType, PolicyDecision, PolicyName, Route};

fn is_high_risk_local_only(t: EntityType) -> bool {
    matches!(
        t,
        EntityType::Ssn | EntityType::Mrn | EntityType::AccountNumber | EntityType::Pin
    )
}

pub fn decide(detection: &DetectionResult, policy_name: PolicyName) -> PolicyDecision {
    let types: HashSet<EntityType> = detection.entities.iter().map(|e| e.kind).collect();

    if matches!(policy_name, PolicyName::HipaaLogging) && detection.has_sensitive() {
        return PolicyDecision {
            route: Route::MaskedSend,
            policy: policy_name,
            rationale: "Strict logging policy: any sensitive data must be masked before traversal."
                .to_string(),
        };
    }

    // GDPR: mask all personal identifiers, allow non-personal context.
    if matches!(policy_name, PolicyName::GdprBase) && detection.has_sensitive() {
        return PolicyDecision {
            route: Route::MaskedSend,
            policy: policy_name,
            rationale: "GDPR base: all personal data must be masked before forwarding.".to_string(),
        };
    }

    if matches!(policy_name, PolicyName::HipaaClinical) {
        if types.iter().any(|&t| is_high_risk_local_only(t)) {
            return PolicyDecision {
                route: Route::LocalOnly,
                policy: policy_name,
                rationale:
                    "Direct identifiers (SSN/MRN) must stay on-device under clinical policy."
                        .to_string(),
            };
        }
        if types.contains(&EntityType::HealthContext) && !detection.entities.is_empty() {
            return PolicyDecision {
                route: Route::MaskedSend,
                policy: policy_name,
                rationale:
                    "Clinical context with identifiers → mask identifiers, keep medical context."
                        .to_string(),
            };
        }
    }

    if types.iter().any(|&t| is_high_risk_local_only(t)) {
        let mut hits: Vec<&'static str> = types
            .iter()
            .filter(|t| is_high_risk_local_only(**t))
            .map(|t| t.as_str())
            .collect();
        hits.sort();
        return PolicyDecision {
            route: Route::LocalOnly,
            policy: policy_name,
            rationale: format!("High-risk identifiers detected: {:?}", hits),
        };
    }

    if detection.has_sensitive() {
        let mut sensitive: Vec<&'static str> = detection
            .entities
            .iter()
            .filter(|e| !matches!(e.kind, EntityType::HealthContext))
            .map(|e| e.kind.as_str())
            .collect();
        sensitive.sort();
        sensitive.dedup();
        return PolicyDecision {
            route: Route::MaskedSend,
            policy: policy_name,
            rationale: format!("Sensitive entities present: {:?}", sensitive),
        };
    }

    PolicyDecision {
        route: Route::SafeToSend,
        policy: policy_name,
        rationale: "No sensitive entities detected.".to_string(),
    }
}
